src/transformers/data/data_collator.py

```diff
@@ -134,3 +134,127 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
 
         # The rest of the time (10% of the time) we keep the masked input tokens unchanged
         return inputs, labels
+
+
+@dataclass
+class DataCollatorForPermutationLanguageModeling:
+    """
+    Data collator used for permutation language modeling.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for permutation language modeling with procedures specific to XLNet
+    """
+
+    tokenizer: PreTrainedTokenizer
+    plm_probability: float = 1 / 6
+    max_span_length: int = 5  # maximum length of a span of masked tokens
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+        return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
+
+    def _tensorize_batch(self, examples: List[torch.Tensor]) -> torch.Tensor:
+        length_of_first = examples[0].size(0)
+        are_tensors_same_length = all(x.size(0) == length_of_first for x in examples)
+        if are_tensors_same_length:
+            return torch.stack(examples, dim=0)
+        else:
+            if self.tokenizer._pad_token is None:
+                raise ValueError(
+                    "You are attempting to pad samples but the tokenizer you are using"
+                    f" ({self.tokenizer.__class__.__name__}) does not have one."
+                )
+            return pad_sequence(examples, batch_first=True, padding_value=self.tokenizer.pad_token_id)
+
+    def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        The masked tokens to be predicted for a particular sequence are determined by the following algorithm:
+            0. Start from the beginning of the sequence by setting `cur_len = 0` (number of tokens processed so far).
```

Comment:
Here and in all docstrings, please use sphinx restructured text syntax (which is different from Markdown): putting an object between backticks just render it in italics. I think you want double-backticks or a syntax like ``obj:`object` ``.
So here either ` ``cur_len = 0`` ` or ``obj:`cur_len=0` `` and same in the rest of this docstring.

Reply:
@sgugger Yes, I think changing it to double backticks should work. I noticed that the docstring in `default_data_collator` also uses single backticks - would you like me to change that as well? Also, is this formatting required in ordinary comments?

Reply:
This is for proper rendering on the online doc, so it does not apply to ordinary comments. Yes please for `default_data_collator`, I haven't gotten to fix this module for the docs yet (not even sure it's documented online yet to be honest).

Reply:
Okay. I've left the single backticks in for the comments to help distinguish variable names and so on from regular text. Let me know if you'd like me to remove those as well.

Reply:
LGTM, thanks!